### ISSUE TITLE
Clustering & prediction on data with uncertainties

### ISSUE THREAD

**Issue by leenderthayen** [scikit-learn-contrib/hdbscan#186]
Hello all,

I'm a theoretical physicist, and have been using hdbscan for a while now. Needless to say, I've been very thankful to the makers for this great piece of software. I have question/request, however.

The data I've been clustering carry uncertainties in some dimensions. For the sake of simplicity, we can assume the uncertainties are gaussian. While clustering with the central values, neglecting any uncertainties, results in physically meaningful clusters, my ultimate interest lies in the prediction of new data. The question is now if it is possible to use the uncertainties with the prediction algorithm. There seems to be (almost) no literature on this, or I have been searching in the wrong directions.

I can imagine defining a custom metric as the euclidian distance normalized to the uncertainty of the individual data points added in quadrature, but I don't see how this works to predict new points. A very brute-force method of performing the clustering and prediction on all data points sampled within their uncertainties will be prohibitively expensive. I can imagine some fiddling with the final posterior could do the trick, but I'm not sure yet if this is the correct way to proceed.

Does anyone have an idea on how to do this, or pointers to some literature?

Thanks!

**Comment by lmcinnes** [comment]
I admit that this is rather tricky. I believe I understand your problem, but I will definitely have to think for a bit about the right way to go about trying to find some solution. It certainly isn't obvious. I'll try to get back to you soon with at least some ideas, or references to potential approaches.

**Comment by leenderthayen** [comment]
Thanks, any help would be very much appreciated! 

Intuitively, I would imagine constructing a weighted gaussian kernel density estimation of the cluster, combined with some chi-squared measure of the proposed point belonging to the resulting pdf would seem appropriate, but I'm not at all sure. I have been going through the 'How soft clustering for HDBSCAN works', but I don't know enough to see how and if it would fit into either the distance or outlier score.

**Comment by lmcinnes** [comment]
SO I ran across this: http://ieeexplore.ieee.org/abstract/document/8309418/ which deals with uncertainty in the metric. I believe you could convert uncertainty in measurement into uncertainty in the metric, and potentially apply some of the approaches there. The conservative thing to do is take the maxiumum value in the uncertainty range on distance for any pairwise distance and call that your quasi-metric, and then cluster. This might work with pre-computed metrics in hdbscan as is, but would require some work for anything else. Let me know if this looks like a promising direction, or if it isn't addressing your concerns.

**Comment by LemonPi** [comment]
I am also very interested in this problem: each of my data points has a covariance around it.
My current assumption is that each point is a Gaussian (point value is the mean and the covariance is the covariance).
I'm trying out 2 approaches for computing distance, 
1) expected Euclidean distance between independent Gaussian random variables: 
![image](https://user-images.githubusercontent.com/5508542/57588052-71a86800-74dc-11e9-87b1-90b3c9bbf145.png)

2) statistical distance between distributions (each point is a distribution) using one of the common metrics:
- Wasserstein (earth mover)
- Hellinger
- Bhattacharyya

I'm implementing these as a custom metric function to pass into hdbscan (they are true metrics) by squeezing the covariances into the data point vector and unsqueezing them inside the metric calculation. Is this the best current way for using hdbscan with covariance information? I'll post some results on how each approach worked later.

**Comment by lmcinnes** [comment]
That all sounds like a reasonable approach. I would think Wasserstein_2 distance would be a good candidate if it were computationally tractable. I look forward to any restults you can share.

**Comment by LemonPi** [comment]
Result post here, but before then an explanation of what I'm clustering. I have a toy 2-D base system where each point is associated with some higher dimensional parameters and a 1-D feature (distance to origin). Call the concatenation of the parameters and the feature as Z. We have covariances for each z, and we're clustering all the z. So we're actually clustering N-D vectors, but can visualize the decision boundary in the base 2-D system. These decision boundaries are made with nearest neighbour classifiers on the feature, so you'll see a bunch of rings.

For comparison I run my current modified kmeans (distance to means using the Mahalanobis metric with the covariance matrix of that data point):
![f_nn_db2](https://user-images.githubusercontent.com/5508542/57956157-11436d00-78c6-11e9-8ae3-6cff6e3d07dd.png)

The ground truth should have 2 clusters, one in the hatched region and one outside, so kmeans is able to find this exactly.

The following are decision boundaries from HDBSCAN using the different metrics and minimum 15 points per cluster. Points classified as noise is filled with black and the classified noise region is filled with red.

Expected Euclidean distance:
![ExpectedEuclideanMetric_cluster](https://user-images.githubusercontent.com/5508542/57957068-0e964700-78c9-11e9-9bb0-9b286c19d9b1.png)
![ExpectedEuclideanMetric_tree](https://user-images.githubusercontent.com/5508542/57957090-266dcb00-78c9-11e9-9a07-35cbc9ec6e36.png)

What's strange here is that the points in the center are all classified as noise. Why are they too far from each other...?

Wasserstein:
![WassersteinMetric_cluster](https://user-images.githubusercontent.com/5508542/57957160-616ffe80-78c9-11e9-94e5-07df8f8338d3.png)
![WassersteinMetric_tree](https://user-images.githubusercontent.com/5508542/57957183-7056b100-78c9-11e9-806d-5468777ecbdd.png)

Weird that the distance of the center points to the outer parts inside the donut is too far for them to be considered a single cluster.

Hellinger:
![HellingerMetric_cluster](https://user-images.githubusercontent.com/5508542/57957249-a85df400-78c9-11e9-9a65-2c21c369bbdb.png)
![HellingerMetric_tree](https://user-images.githubusercontent.com/5508542/57957253-aac04e00-78c9-11e9-8e9c-38d973aac64b.png)

Saturation at 0 and 1 doesn't work that well here.

Bhattacharyya:
![BhattacharyyaMetric_cluster](https://user-images.githubusercontent.com/5508542/57957290-c9bee000-78c9-11e9-9bbe-3d7567f88f3b.png)
![BhattacharyyaMetric_tree](https://user-images.githubusercontent.com/5508542/57957291-cc213a00-78c9-11e9-92d1-4d15b9ae43dd.png)

This one seems to have properties that I want, but is still classifying some of the ones that shouldn't be noise as noise. The points near boundaries being noise is acceptable, but there are some that are safely outside the donut. 

Another issue is performance. Running the Bhattacharyya clustering for 200 points takes 0.5 seconds whereas using the default Euclidean metric takes 0.005 seconds. Does HDBSCAN know when the input metric is a metric or not? (actually Bhattacharyya doesn't have triangle inequality)

**Comment by lmcinnes** [comment]
Thanks for this -- I admit to having some difficulty interpreting all the plots. I think some of the issues are due to the (relatively) low number of samples. With more data I suspect hdbscan may perform better (in terms of quality of results). In terms of performance: the current implementation has a (largeish) set of predefined metric functions that have a fast path since they are known to be metrics and supported by KD-trees. As soon as you are outside that set of metrics then a different code-path is used and it is necessarily much slower. Obviously Euclidean is the default and is supported by the faster code path, while all the custom metrics you have here take the slower approach.

**Comment by MilesCranmer** [comment]
Just adding to this: Wasserstein_2 is also pretty simple for diagonal covariance matrices:
![image](https://user-images.githubusercontent.com/7593028/79576028-a92a7780-8090-11ea-8691-e7de39199d07.png)

It would be great if this could be added somehow!!

**Comment by lmcinnes** [comment]
Thanks @MilesCranmer . The big catch in this codebase is that I am relying on sklearn's KD-Trees and BallTrees for acceleration. That means that until they implement a metric I don't really have access to it for fast computation approaches. In the longer term I would hope to re-implement hdbscan using tricks like NNDescent for k-neighbor graph computation. In that case we can certainly have all manner of metrics, including this, a Wassterstein/Kantorvich distance between general multivariate Gaussians, and even a general Wasserstein/Kantorovich distance between discrete distributions over a metric space. All of that, however, is unlikely to happen soon.

**Comment by MilesCranmer** [comment]
Thanks for the quick response as always @lmcinnes! Sorry for my late response - I think my GH notifications are broken again; will fix it now.
Ah, I understand this issue now - I'll check with sklearn about including this metric. By the way, sorry if I missed this in my reading of the hdbscan docs, but could one use a separate distance metric for KD-Trees and then hdbscan? E.g., get a list of closest neighbors using Euclidean distance, and then use Wasserstein to calculate the mutual reachability distance?
